File: crates/sl-frontend/src/lib/managers/instances.ts
import { reactive } from 'vue';
import { InstanceMetadata, ModLoader } from '@/types/instances';
import { invoke } from '@tauri-apps/api/core';

export const instancesManager = reactive({
    loading: false,
    instances: [] as InstanceMetadata[],

    add: (instance: InstanceMetadata) => {
        instancesManager.instances.push(instance)
    },

    get: (name: string) => {
        return instancesManager.instances.find(instance => instance.name === name);
    },

    delete: (name: string) => {
        instancesManager.instances = instancesManager.instances.filter(instance => instance.name != name);
    },

    refresh: async () => {
        instancesManager.loading = true;
        instancesManager.instances = await instancesManager.get_all_instances();
        instancesManager.loading = false;
    },

    init: async () => {
        if (instancesManager.instances.length === 0) {
            await instancesManager.refresh();
        }
    },

    get_all_instances: async () => {
        try {
            const result = await invoke<InstanceMetadata[]>("get_all_instances");
            return result as InstanceMetadata[];
        } catch (error) {
            console.error(`get_all_instances error: ${error}`);
            return []
        }
    },

    create_instance: async (
        name: string,
        game_version: string,
        mod_loader: ModLoader,
        mod_loader_version?: string
    ) => {
        try {
            const new_instance = await invoke<InstanceMetadata>("create_instance", {
                instanceName: name,
                gameVersion: game_version,
                loader: mod_loader,
                loaderVersion: mod_loader_version
            })

            instancesManager.add(new_instance)
        } catch (error) {
            console.error(`create_instance error: ${error}`)
        }
    },

    launch_instance: async (name: string) => {
        try {
            await invoke("launch_instance", { name: name })
        } catch (error) {
            console.error(`launch_instance error: ${error}`)
        }
    },

    delete_instance: async (name: string) => {
        try {
            await invoke("delete_instance", { instanceName: name });
            instancesManager.delete(name)
        } catch (error) {
            console.error(`delete_instance error: ${error}`)
        }
    },

    // edit_instance: async (name: string, new_mc_version?: string, new_modloader_version?: string) => {
    //     try {

    //     } catch (error) {
    //         console.error(`edit_instance error: ${error}`)
    //     }
    // }
});

await instancesManager.init()

File: crates/sl-frontend/src-tauri/src/command/instances.rs
use serde::{Deserialize, Serialize};
use sl_core::{
    environment::LauncherEnv,
    instances::instance_metadata::{InstanceMetadata, ModLoader},
};
use tauri::State;
use tokio::sync::RwLock;

#[derive(Debug, Deserialize, Serialize)]
pub struct Instance {
    #[serde(flatten)]
    metadata: InstanceMetadata,
    icon: Option<Vec<u8>>,
}

#[tauri::command]
pub async fn create_instance(
    instance_name: String,
    game_version: &str,
    loader: ModLoader,
    loader_version: Option<String>,
    launcher_env: State<'_, RwLock<LauncherEnv>>,
) -> Result<InstanceMetadata, String> {
    let env = launcher_env.read().await;

    let instance = env.instances()
        .create_instance(instance_name, &game_version, loader, loader_version)
        .await.map_err(|e| e.to_string())?;
    
    Ok(instance)
}

#[tauri::command]
pub async fn get_instance(
    name: &str,
    launcher_env: State<'_, RwLock<LauncherEnv>>,
) -> Result<Instance, String> {
    let env = launcher_env.read().await;

    let instance_metadata = env
        .instances()
        .get_existing(name)
        .map_err(|e| e.to_string())?
        .0;
    let icon = instance_metadata.get_instance_icon(&env.instances()).await;

    Ok(Instance {
        metadata: instance_metadata,
        icon: icon,
    })
}

#[tauri::command]
pub async fn get_all_instances(
    launcher_env: State<'_, RwLock<LauncherEnv>>,
) -> Result<Vec<Instance>, String> {
    let mut instances = Vec::new();
    let env = launcher_env.read().await;

    let instance_metadatas = env
        .instances()
        .get_all_instances()
        .await
        .map_err(|e| e.to_string())?;

    for metadata in instance_metadatas {
        let icon = metadata.get_instance_icon(&env.instances()).await;
        instances.push(Instance {
            metadata: metadata,
            icon: icon,
        });
    }

    Ok(instances)
}

#[tauri::command]
pub async fn launch_instance(
    name: &str,
    launcher_env: State<'_, RwLock<LauncherEnv>>,
) -> Result<(), String> {
    let env = launcher_env.read().await;

    let instance_metadata = env
        .instances()
        .get_existing(name)
        .map_err(|e| e.to_string())?
        .0;

    instance_metadata
        .load_init(&env.instances())
        .await
        .map_err(|e| e.to_string())?
        .execute()
        .await
        .map_err(|e| e.to_string())?;

    Ok(())
}
